Refine data documentation in /developers/docs/data-and-analytics/ (#13352)
ethereum/ethereum-org-website: public/content/developers/docs/data-and-analytics/index.md

Reviewer: Seeing a word count around 1800

https://ethereum.org/en/contributing/style-guide/#best-practices

Generally try to keep things closer to 1000, 1500 max... perhaps we could split the content of this page into two separate pages?

---
title: Data and Analytics
description: Understanding and navigating the crypto data stack
lang: en
---

## Introduction {#Introduction}

As you get into crypto, you're likely to see stats, charts, and APIs across dozens of different websites (and github repos). This is a stark contrast from traditional tech or finance where there is just one trusted source (i.e. Bloomberg). This is due to the open data nature of the industry, where all data can be freely downloaded from client nodes.

> However, that data is stored in hex/binary form and is hardly legible or usable. As an ecosystem, we've had to work together to share queries, tools, tables, and labor to make crypto data accessible for everyone. Each year, we see giant leaps in how we can combine onchain/offchain data in analytics and applications. 

Reviewer: ah, see it's covered below, that'll do

> In this overview page, we'll cover all the parts of the crypto data stack as they stand today (last updated in 2024).

Reviewer: @andrewhong5297 Certainly apologize for delays here... anything critical that has developed since that you'd like to add now that we're in 2025?

> ## Crypto Data Landscape

Reviewer: (Custom header ids needed for all headers—related to our translations)

```suggestion
## Crypto Data Landscape {#crypto-data-landscape}
```

> Below is a diagram of the five parts of the crypto data stack, with some top example product logos called out. You can learn more about these products in the [tool descriptions page](public/content/developers/docs/data-and-analytics/tool-descriptions.md)

Reviewer: ```suggestion
Below is a diagram of the five parts of the crypto data stack, with some top example product logos called out. You can learn more about these products in the [tool descriptions page](/developers/docs/data-and-analytics/tool-descriptions/)
```

And similarly, that page should be at tool-descriptions/index.md, not tool-descriptions.md; we'll get it moved

> ![](/public/images/data/data_landscape.jpg)

Reviewer: (Alt text)

```suggestion
![Diagram showing the web3 data stack as represented in 2024](/public/images/data/data_landscape.jpg)
```

> #### web2 counterparts

Reviewer: (Important not to skip heading levels)

```suggestion
### Web2 counterparts {#web2-counterparts)
```

If you are new to web3 and this looks confusing to you, think of these web2 parallels:

- indexing: amplitude, stripe, general logging services
- explore: grafana, datadog
- query: metabase, hex
- define and store: dbt, snowflake, databricks

> - …and back onchain: this is just reverse ETL

Reviewer: Any proper names/brands should be capitalized here

> ### Index: Read raw data from the blockchain

Reviewer: ```suggestion
### Index: Read raw data from the blockchain {#index}
```

Blockchains run with nodes. Nodes run “client” code - which are repos that have implemented the EVM in some fashion. These clients have a set of implemented RPCs (API endpoints), some of which are standard and some of which are custom to support better/faster data querying. Data from RPC endpoints are direct reflections of the blockchain state - there are no external party changes or transformations here. Nodes-as-a-service providers run tons of nodes and offer them collectively through an API product, creating a stable data service.

We’ve seen a ton of change in the indexing layer over the last year, with two new subcategories emerging:

> - Forks-as-a-Service

Reviewer: ```suggestion
**Forks-as-a-Service**
```

>   - Fork any contract and add events and calculations, and then pull this data from a new “forked” RPC/data service.
>   - Some of the main providers for this are [shadow.xyz](https://www.shadow.xyz/), [ghostlogs](https://ghostlogs.xyz/), and [smlXL](https://smlxl.io/).
>   - ilemi gave his thoughts on [shortcomings and difficulties](https://twitter.com/andrewhong5297/status/1732230186966413484) on this approach.

Reviewer: ```suggestion
- Fork any contract and add events and calculations, and then pull this data from a new “forked” RPC/data service.
- Some of the main providers for this are [shadow.xyz](https://www.shadow.xyz/), [ghostlogs](https://ghostlogs.xyz/), and [smlXL](https://smlxl.io/).
- ilemi gave his thoughts on [shortcomings and difficulties](https://twitter.com/andrewhong5297/status/1732230186966413484) on this approach.
```

> - Rollups-as-a-Service (RaaS):

Reviewer: ```suggestion
**Rollups-as-a-Service (RaaS):**
```

>   - The big theme of the year has been rollups, with Coinbase kicking it off by launching a rollup (Base) on the Optimism Stack (OP Stack) earlier this year.

Reviewer: As you can see, time passes fast... using terms like "earlier this year" quickly become dated. Ideally we rephrase to keep content a bit more future-proof, although I do appreciate the note at the top mentioning the timeframe we're discussing.

>   - The big theme of the year has been rollups, with Coinbase kicking it off by launching a rollup (Base) on the Optimism Stack (OP Stack) earlier this year.
>   - Teams are building products specifically for running the nodes and sequencer(s) for your own Rollup. We’ve already seen dozens of rollups launch.
>   - New startups like [Conduit](https://conduit.xyz/), [Caldera](https://caldera.xyz/), and [Astria](https://www.astria.org/) are offering full stack rollup services. Quicknode and Alchemy have launched similar RaaS offerings.

Reviewer: ```suggestion
- The big theme of the year has been rollups, with Coinbase kicking it off by launching a rollup (Base) on the Optimism Stack (OP Stack) earlier this year.
- Teams are building products specifically for running the nodes and sequencer(s) for your own Rollup. We’ve already seen dozens of rollups launch.
- New startups like [Conduit](https://conduit.xyz/), [Caldera](https://caldera.xyz/), and [Astria](https://www.astria.org/) are offering full stack rollup services. Quicknode and Alchemy have launched similar RaaS offerings.
```

> Alchemy and Quicknode have expanded further into crypto native infra and data engineering infra. Alchemy launched [account abstraction](https://www.alchemy.com/account-abstraction-infrastructure) and [subgraph services](https://www.alchemy.com/subgraphs). Quicknode has been busy with [alerts](https://www.quicknode.com/quickalerts), [data streaming](https://www.quicknode.com/quickstreams), and [rollup services](https://blog.quicknode.com/introducing-quicknode-custom-chains).

Reviewer: Few updated links

```suggestion
Alchemy and Quicknode have expanded further into crypto native infra and data engineering infra. Alchemy launched [account abstraction](https://www.alchemy.com/bundler) and [subgraph services](https://www.alchemy.com/subgraphs). Quicknode has been busy with [alerts](https://www.quicknode.com/quickalerts), [data streaming](https://www.quicknode.com/stream), and [rollup services](https://blog.quicknode.com/introducing-quicknode-custom-chains).
```

> We should see our first “intents” clients/services soon. Intents are a part of the modular stack, and are essentially transactions handled outside the mempool that have extra preferences attached. UniswapX and Cowswap both operate limit order intent pools, and should both release clients within the year. Account abstraction bundlers like [stackup](https://www.stackup.sh/) and [biconomy](https://www.biconomy.io/) should venture into intents as well. It’s unclear if data providers like Alchemy will index these “intents” clients, or if it will be like MEV where we have specialized providers like [Blocknative](https://www.blocknative.com/) and [Bloxroute](https://bloxroute.com/).

Reviewer: ```suggestion
We should see our first “intents” clients/services soon. Intents are a part of the modular stack, and are essentially transactions handled outside the mempool that have extra preferences attached. UniswapX and Cowswap both operate limit order intent pools, and should both release clients within the year. Account abstraction bundlers like [stackup](https://www.stackup.fi/) and [biconomy](https://www.biconomy.io/) should venture into intents as well. It’s unclear if data providers like Alchemy will index these “intents” clients, or if it will be like MEV where we have specialized providers like [Blocknative](https://www.blocknative.com/) and [Bloxroute](https://bloxroute.com/).
```

Another up-and-coming type of provider is the “all-in-one” service, which combines indexing, querying, and defining. There are a few products here such as [indexing.co](https://www.indexing.co/) and [spec.dev](https://spec.dev/) - they are not included them in the landscape since they are still nascent.

> ### Explore: Quickly look into addresses, transactions, protocols, and chains

Reviewer: ```suggestion
### Explore: Quickly look into addresses, transactions, protocols, and chains {#explore}
```

It’s common to jump between well-crafted data dashboards and blockchain explorers iteratively, to help you identify a trend or build out a cohesive data pipeline.

> Outside of etherscan, we now have this plethora of explorers to choose from:

Reviewer: ```suggestion
Outside of Etherscan, we now have this plethora of explorers to choose from:
```

- More intuitive explorers like [parsec](https://parsec.fi/), [arkham](https://platform.arkhamintelligence.com/), [onceupon](https://www.onceupon.gg/home) showcasing more metadata and charts for transactions and addresses.
- In depth explorers like [evm.storage](https://evm.storage/eth/18906826/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48#map) showcasing storage slots and memory stacks
- Cross-chain explorers like [Routescan](https://routescan.io/) (superscan), [Dora](https://www.ondora.xyz/), and [Onceupon](https://www.onceupon.gg/home)
- ZK proof explorers like [modulus](https://explorer.modulus.xyz/batch/157/inference/61974), [succinct](https://alpha.succinct.xyz/), and [axiom](https://explorer.axiom.xyz/v1/mainnet).
- Bridge explorers like [socketscan](https://www.socketscan.io/) and [wormhole](https://wormhole.com/explorer/)
- MEV explorers like [Eigenphi](https://eigenphi.io/mev/ethereum/txr) for transactions or [mevboost.pics](https://mevboost.pics/) for bundles and [beaconcha.in](https://beaconcha.in/) for blocks

> - Nansen launched a 2.0 of their token and wallet tracking product, with cool new features like “smart segments”

Reviewer: Noting we do already have a dedicated page for block explorers nested in this section: https://ethereum.org/en/developers/docs/data-and-analytics/block-explorers/

It's hard enough to maintain one list of these; I'd be hesitant to add a list with the same purpose to this page. cc: @konopkja

The dashboard layer hasn’t changed much. It’s still the wild west here. If you spend a day on Twitter, you’ll see charts from dozens of different platforms covering similar data but all with slight differences or twists. Verification and attribution are becoming a bigger issue, especially now with the large growth in both teams and chains.

> Marketing specific address explorers are on the rise. Teams like spindl and bello will lead the way here. Cross-chain explorers (and pre-chain like MEV/intents) will see expansion in development.

Reviewer: Capitalize team names? Only leave lower if they're branding indicates as such... kinda like `ethereum.org` =)

> Across platforms, wallets are still poorly labelled and tracked, and it’s getting worse with intents/account abstraction now. We don’t mean static labels like “Coinbase” but instead more dynamic ones like “Experienced Contract Deployer”. Some teams are trying to tackle this such as [walletlabels](https://walletlabels.xyz/), [onceupon context](https://github.com/Once-Upon/context), and [syve.ai](https://www.syve.ai/). It will also improve naturally alongside web3 social, which is growing mainly on Farcaster.

Reviewer: https://walletlabels.xyz
<img width="454" height="183" alt="image" src="https://github.com/user-attachments/assets/636d506a-cbf8-404f-8115-eca96041548a" />

<img width="605" height="509" alt="image" src="https://github.com/user-attachments/assets/60c5193e-d321-4b9c-b528-a0b7564ba48b" />

sad panda..

```suggestion
Across platforms, wallets are still poorly labelled and tracked, and it’s getting worse with intents/account abstraction now. We don’t mean static labels like “Coinbase” but instead more dynamic ones like “Experienced Contract Deployer”. Some teams are trying to tackle this, such as [syve.ai](https://www.syve.ai/). It will also improve naturally alongside web3 social, which is growing mainly on [Farcaster](https://farcaster.xyz/).
```

> ### Query: Raw, decoded, and abstracted data that can be queried

Reviewer: ```suggestion
### Query: Raw, decoded, and abstracted data that can be queried {#query}
```

Most SQL query engines are cloud-based, so you can use an in-browser IDE to query against raw and aggregated data (like nf.trades/dex.trades). These also allow for definition of great tables such as NFT wash trading filters. All these products come with their own APIs to grab results from your queries.

GraphQL APIs here let you define your own schemas (in typescript or SQL) and then generates a graphQL endpoint by running the full blockchain history through your schema.

For predefined APIs (where you query prebuilt schemas), there are a ton of niche data providers that are not included in my chart. Data providers covering domains like mempool, nft, governance, orderbook, prices, and more.

Holistically, every platform has got a lot more efficient with their infra (meaning your queries run faster). Most platforms explored advanced methods of getting data out like odbc connectors, data streams, s3 parquet sharing, bigquery/snowflake direct transfers, etc.

> Recent changes to existing products:

Reviewer: (Again, careful with "recent" language)

> 1. Query engines like [Dune](https://dune.com/) and [Flipside](https://flipsidecrypto.xyz/) have accepted there is more data than can possibly be ingested in custom data pipelines, and have launched products that allow the user to bring in that data instead. Flipside launched livequery (query an API in SQL) and Dune launched uploads/syncs (upload csv or api, or sync your database to Dune directly).

Reviewer: ```suggestion
1. Query engines like [Dune](https://dune.com/) and [Flipside](https://flipsidecrypto.xyz/) have accepted there is more data than can possibly be ingested in custom data pipelines, and have launched products that allow the user to bring in that data instead. Flipside launched livequery (query an API in SQL) and Dune launched uploads/syncs (upload csv or api, or sync your database to Dune directly).

```

> 2. Our favorite decentralized data child, [The Graph](https://thegraph.com/), has had to really beef up their infra to not lose market share to centralized subgraph players like goldsky and satsuma (alchemy). They’ve partnered closely with [StreamingFast](https://www.streamingfast.io/), separating out “reader” and “relayer” of data and also introducing [substreams](https://thegraph.com/docs/en/substreams/) which allow you to write rust based subgraphs across chains.

Reviewer: "Our favorite decentralized data child" -- I know it's colloquial, but may re-word this for a platform like ethereum.org

No provider here is truly set up for the rollup world yet, either in terms of scaling ingestion or fixing cross-chain schemas/transformations. And by not ready, we mean not ready for the case of 500 rollups launching in a week. Dune has launched a [rollup ingestion product](https://dune.com/product/dune-catalyst) to start making this easier, especially if you use an existing RaaS provider like Conduit.

LLM query/dashboard products like Dune AI will start to gain stronger traction in certain domains, such as wallet analysis or token analysis. Labels datasets will play a strong part in enabling this.

> ### Define and Store: Create and store aggregations that rely on heavy data transformations

Reviewer: ```suggestion
### Define and Store: Create and store aggregations that rely on heavy data transformations {#define-and-store}
```

Raw data is great, but to get to better metrics, you need to be able to standardize and aggregate data across contracts of different protocols. Once you have aggregations, you can create new metrics and labels that enhance everyone's analysis. We've have only included products that have active contribution from both inside and outside the platform’s team, and are publicly accessible.

The collaborative layer of data definition has not really evolved over the last year. Product teams and engineering are barely keeping up as is. To give a sense of growth rate here, in the month of December 2023:

- DeFiLlama adapters saw [230 pull requests from 150 contributors across 559 files](https://github.com/DefiLlama/DefiLlama-Adapters/pulse/monthly) out of about 3700 total files.
- Dune spellbook saw [127 pull requests from 42 contributors across 779 files](https://github.com/duneanalytics/spellbook/pulse/monthly) out of about 3700 total files.

> You can probably consider every 2-3 files to be a new protocol/table. So 200+ tables shifting around every month, and that number is only increasing with new chains and startups. If you’ve never worked in GitHub or data tables before, let me tell you that’s a ton of work to manage.

Reviewer: Similar to above... "let me tell you that’s a ton of work to manage." would reword

Some general ethereum.org style guide tips: https://ethereum.org/en/contributing/style-guide/#best-practices

> ### Back Onchain: Putting the data back into contracts using Zero Knowledge (ZK) tech

Reviewer: ```suggestion
### Back onchain: Putting the data back into contracts using zero knowledge (ZK) tech {#back-onchain}
```
